modules/common/identity/dynamic-hostname.nix

```
@@ -0,0 +1,141 @@
+# SPDX-FileCopyrightText: 2022-2026 TII (SSRC) and the Ghaf contributors
+# SPDX-License-Identifier: Apache-2.0
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.ghaf.identity.dynamicHostName;
+
+  computeScript = pkgs.writeShellScript "ghaf-compute-hostname" ''
+    set -euo pipefail
+
+    PATH=${
+      lib.makeBinPath [
+        pkgs.coreutils
+        pkgs.iproute2
+        pkgs.gawk
+        pkgs.gnugrep
+      ]
+    }:$PATH
+
+    prefix=${lib.escapeShellArg cfg.prefix}
+    outDir=${lib.escapeShellArg (toString cfg.outputDir)}
+    shareDir=${lib.escapeShellArg (toString cfg.shareDir)}
+    digits=${toString cfg.digits}
+
+    mkdir -p "$outDir" "$shareDir"
+
+    read_dmi() {
+      for p in /sys/class/dmi/id/product_serial /sys/class/dmi/id/product_uuid; do
+        if [ -r "$p" ]; then
+          v=$(tr -cd '[:alnum:]' <"$p")
+          if [ -n "$v" ]; then
+            echo "$v"; return 0
+          fi
+        fi
+      done
+      return 1
+    }
+
+    read_mac() {
+      ${pkgs.iproute2}/bin/ip -o link show \
+        | ${pkgs.gawk}/bin/awk -F': ' '$2 !~ /lo/ {print $2}' \
+        | while read -r ifc; do
+            [ -r "/sys/class/net/$ifc/address" ] && cat "/sys/class/net/$ifc/address"
+          done \
+        | ${pkgs.gnugrep}/bin/grep -vi '^00:00:00:00:00:00$' | sort | head -n1
+    }
+
+    key=""
+    key=$(read_dmi || true)
+    if [ -z "$key" ]; then
+      key=$(read_mac || true)
+    fi
+    if [ -z "$key" ] && [ -r /etc/machine-id ]; then
+      key=$(cat /etc/machine-id)
+    fi
+    if [ -z "$key" ]; then
+      key="fallback"
```

Instead of accepting this generic fallback ID, maybe this should be a hard fail during boot? Implication is that someone is trying to boot Ghaf on incompatible hardware.

---

It is of course can be implemented, I'm actually not sure if it is a good idea to break the boot process. @brianmcgillion your opinion?